src/System.Net.Http.Json/src/System/Net/Http/Json/TranscodingReadStream.cs

```diff
@@ -0,0 +1,244 @@
+                    throw new InvalidOperationException();
+            }
+        }
+
+        private async Task ReadInputChars(CancellationToken cancellationToken)
+        {
+            Debug.Assert(_byteBuffer.Array != null);
+            // If we had left-over bytes from a previous read, move it to the start of the buffer and read content in to
+            // the segment that follows.
+            Buffer.BlockCopy(
+                _byteBuffer.Array,
+                _byteBuffer.Offset,
+                _byteBuffer.Array,
+                0,
+                _byteBuffer.Count);
+            int readBytes =
+                await _stream.ReadAsync(_byteBuffer.Array, _byteBuffer.Count, _byteBuffer.Array.Length, cancellationToken).ConfigureAwait(false);
+            _byteBuffer = new ArraySegment<byte>(_byteBuffer.Array, 0, _byteBuffer.Count + readBytes);
+
+            Debug.Assert(_charBuffer.Count == 0, "We should only expect to read more input chars once all buffered content is read");
+
+            _decoder.Convert(_byteBuffer.Array, _byteBuffer.Offset, _byteBuffer.Count, _charBuffer.Array, 0, _charBuffer.Array.Length,
+                flush: readBytes == 0, out int bytesUsed, out int charsUsed, out _);
+
+            Debug.Assert(_charBuffer.Array != null);
+
+            _byteBuffer = _byteBuffer.Slice(bytesUsed);
+            _charBuffer = new ArraySegment<char>(_charBuffer.Array, 0, charsUsed);
+        }
+
+        private static void ThrowArgumentOutOfRangeException(byte[] buffer, int offset, int count)
+        {
+            if (count < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count));
+            }
+
+            if (offset < 0 || offset >= buffer.Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(offset));
+            }
+
+            if (buffer.Length - offset < count)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count));
+            }
+        }
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override void SetLength(long value)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            throw new NotSupportedException();
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (!_disposed)
+            {
+                _disposed = true;
+
+                Debug.Assert(_charBuffer.Array != null);
+                ArrayPool<char>.Shared.Return(_charBuffer.Array);
+
+                Debug.Assert(_byteBuffer.Array != null);
+                ArrayPool<byte>.Shared.Return(_byteBuffer.Array);
+
+                Debug.Assert(_overflowBuffer.Array != null);
+                ArrayPool<byte>.Shared.Return(_overflowBuffer.Array);
```

Comment:
This should null out the various buffer fields.  Otherwise, misuse of this stream could end up accessing arrays that have already been returned to the pool.